Added optional redis layer

MrNaif2018 opened the pull request:
This pull requests adds RedisLayer class.
As starlette endpoint was hardcoded to use default one, now it is possible to select one by setting class attribute.
To install redis layer, it is needed to install optional dependency i.e.:
`pip install nejma[redis]`
I think that all functions(except for utility ones) should become asyncronous.
What's the point in having most functions async and only add function sync for example?
This pull request is probably not yet ready to merge as some compatibility changes should be decided.

MrNaif2018 committed:
Added optional redis layer

taoufik07 commented:
Thanks @MrNaif2018 , you're right we should make the `add` method a coroutine.

MrNaif2018 commented:
In this case I guess I should update that in default channel layer and in readme, right?

eliclement commented on nejma/layers.py:
```diff
@@ -81,4 +86,57 @@ def clean_expired(self):
+            f"{self.prefix}_") else f"{group}"
+
+    async def _initialize(self):
+        self.redis: aioredis.Redis = await aioredis.create_redis(self.redis_host)
```
Thanks for this great implementation @MrNaif2018. Would there be any advantage in using
`aioredis.create_redis_pool(self.redis_host)` instead of `aioredis.create_redis(self.redis_host)`. According to aioredis docs this should give 

> Redis client bound to pool of connections (allows auto-reconnecting).

MrNaif2018 replied:
Yes, auto-reconnecting and in general connection pool would be better. Will change that in next commit.

MrNaif2018 committed:
Use redis pool, all public apis are async

MrNaif2018 commented:
@eliclement I have changed `create_redis` to `create_redis_pool.`
Also added flush method which was missing and I made all public apis async(and updated README).

MrNaif2018 commented:
@taoufik07 Is it ready to merge?

MrNaif2018 committed:
Fix for group not being str sometimes

MrNaif2018 commented:
Actually why does name.isindentifier()(`validate_group`) check is added? Does it break something if channel name is just numeric?

taoufik07 commented on nejma/layers.py:
```diff
@@ -73,12 +78,73 @@ def validate_name(self, name):
 
 
+class RedisLayer(ChannelLayer):
+    def __init__(self, redis_host="redis://localhost", prefix="group"):
```
Can you add an assert to check if `aioredis` is installed and some message e.g. `assert aioredis is not None, "aioredis must be installed to use RedisLayer"`

MrNaif2018 replied:
Yeah, I first thought that aioredis defaulting to None would be ok.
But I would use exception, maybe ValueError? or AssertionError, as asserts aren't good in production

MrNaif2018 replied:
Done

MrNaif2018 replied:
Assertion error is raised

taoufik07 commented on nejma/layers.py:
```diff
@@ -73,12 +78,73 @@ def validate_name(self, name):
+        if not self.initialized:
+            await self._initialize()
+        group_name = self.group_prefix(group)
+        for _ in await self.redis.hgetall(group_name):
```
you can use `async for` to iterate over coroutine e.g. `async for _ in self.redis.hgetall...`

MrNaif2018 replied:
Yes, but as far as I remember it doesn't support it, no __aiter__ implemented for hgetall.
I get this error:
![изображение](https://user-images.githubusercontent.com/39452697/63441981-fb03fa80-c43a-11e9-8ec7-cff82a4499ec.png)

MrNaif2018 replied:
Hgetall doesn't support it.

taoufik07 commented on nejma/layers.py:
```diff
@@ -73,12 +78,73 @@ def validate_name(self, name):
+        if not self.initialized:
+            await self._initialize()
+        _, keys = await self.redis.scan(match=self.group_prefix("*"))
+        for key in keys:
```
same here `async for`, you can save a line

can you please do it to other for loops that fells in the same pattern ?

MrNaif2018 replied:
I think only scan-line commands support it.
https://aioredis.readthedocs.io/en/v1.2.0/start.html#python-3-5-async-with-async-for-support

MrNaif2018 replied:
Well, for all scan commands I will use iscan then, which is iterative scan.
But for hgetall it can't be done

MrNaif2018 committed:
Fixes

MrNaif2018 commented:
@taoufik07 I have fixed all the reviewed issues.
But `hgetall` doesn't support `async for` interface.

eliclement commented:
@taoufik07 can this pull request be merged?

MrNaif2018 commented:
@taoufik07 Could this pr get merged, more than two months already passed? If you can't maintain this project, I can probably maintain it. What do you think?

MrNaif2018 committed:
Remove name validation for RedisLayer

eliclement commented:
@taoufik07 are there any code review or suggestions needed or to add to get this merged?

MrNaif2018 commented:
This was long time ago, but if someone needs just something like redis for websocket sync, you can use the following utility functions:

```
async def make_subscriber(name):
    subscriber = await aioredis.create_redis_pool(settings.REDIS_HOST)
    res = await subscriber.subscribe(f"channel:{name}")
    channel = res[0]
    return subscriber, channel
```
```
async def publish_message(channel, message):
    return await settings.redis_pool.publish_json(f"channel:{channel}", message)
```

Where settings.redis_pool can be created like so:

```
loop = asyncio.get_event_loop()
redis_pool = None


async def init_redis():
    global redis_pool
    redis_pool = await aioredis.create_redis_pool(REDIS_HOST)


loop.create_task(init_redis())
```

And in websocket route:
```
@router.websocket_route("/ws/some/endpoint")
class WebsocketTestEndpoint(WebSocketEndpoint):
    subscriber = None

    async def on_connect(self, websocket, **kwargs):
        await websocket.accept()
        self.channel_name = secrets.token_urlsafe(32) # or other id
        self.subscriber, self.channel = await utils.make_subscriber(self.wallet_id)
        settings.loop.create_task(self.poll_subs(websocket))

    async def poll_subs(self, websocket):
        while await self.channel.wait_message():
            msg = await self.channel.get_json()
            await websocket.send_json(msg)

    async def on_disconnect(self, websocket, close_code):
        if self.subscriber:
            await self.subscriber.unsubscribe(f"channel:{self.channel_name}")
```

See [example](https://github.com/bitcartcc/bitcart/blob/8b2c66d61a98f5c2e0328ce7fdde4f22d4ad631b/api/views.py#L623) in my repo

I doubt this PR would get merged, as the project seems unmaintained, so here's a custom solution I use.

eliclement commented:
@MrNaif2018 thank you for this custom solution! In the meanwhile Encode is also working on a nice simple broadcast API (called [broadcaster](https://github.com/encode/broadcaster)) that works with a number of different backend services such as Redis PUB/SUB, Apache Kafka, and Postgres LISTEN/NOTIFY. It is still in devolpment but seems very promising.

MrNaif2018 commented:
Thanks for suggesting broadcaster! When it will get out of alpha and when I drop python 3.6 support I will definitely consider it in my fastapi app! Even if this doesn't get merged this page can serve good for providing different implementation examples.

taoufik07 commented:
Thanks @MrNaif2018 for the amazing work and I'm sorry for not following this PR.

I had some thoughts to redesign **nejma** and decoupled it from the web framework so it can be easy to maintain and fit nicely with the ASGI ecosystem, but I was overwhelmed with my final year report and startup work.

I think that what [broadcaster](https://github.com/encode/broadcaster) did represents partly what I wanted to do in my redesign and all the efforts should go there, even tho I'll try to update my draft and discuss it here just for fun.

taoufik07 reviewed (approved):
Good job, thanks !
